### ISSUE TITLE
Data Explorer Integration

### ISSUE THREAD

**Issue by kevinworthington** [IQSS/dataverse#4249]
With the development of the [Data Explorer](https://github.com/scholarsportal/Dataverse-Data-Explorer), having the option for others to easily integrate it with their Dataverse is the next step.

At present, the "Explorer" button in Dataverse has been hi-jacked to link to the standalone Data Explorer application passing URL parameters containing the required details. 
There has also been the inclusion of a [ReadFile.java](https://github.com/scholarsportal/SP-dataverse/blob/SP_v4.6.1.1/src/main/java/edu/harvard/iq/dataverse/export/ReadFile.java) file to overcome the "net::ERR_CONTENT_LENGTH_MISMATCH error" when loading the Tab file for tabular display.
Another piece of the puzzle involves security/privacy controls which would limit features based on those desired by the content publisher. This last item relates to ticket [4230](https://github.com/IQSS/dataverse/issues/4230)

**Comment by pdurbin** [comment]
> There has also been the inclusion of a ReadFile.java file to overcome the "net::ERR_CONTENT_LENGTH_MISMATCH error" when loading the Tab file for tabular display.

This is the Servlet I mentioned during Sprint Planning today. Data Explorer won't work without it so Module Explore (#3657) won't be enough for support Data Explorer unless we add the Servlet at some point. Or figure out a workaround?

@kevinworthington I'm wondering if we should have a dedicated issue for the net::ERR_CONTENT_LENGTH_MISMATCH error. We're trying to work in small chunks. Are the steps to reproduce this straightforward? Can you reproduce this error at https://demo.dataverse.org ? Again, a separate issue might be better.

> Another piece of the puzzle involves security/privacy controls which would limit features based on those desired by the content publisher.

I'm confused by this. Is there something you need on the Dataverse side for this? Maybe this could be a "phase 2" effort? Just get is working with public files to start, perhaps?

**Comment by pdurbin** [comment]
@kevinworthington just a heads up that over at http://irclog.iq.harvard.edu/dataverse/2017-11-17#i_60419 I was encouraging you to look at pull request #4243 but today we decided to close that pull request an open pull request #4298. In short, there's a new branch to look at if you want to try to help us work on improving the parser for the "Modular Explorer" issue at #3657 (though I see at https://waffle.io/IQSS/dataverse that it's now a candidate for the next sprint). You might want to ping me at http://chat.dataverse.org to check in before you do too much coding. I appreciate your willingness to help!

**Comment by kevinworthington** [comment]
Hi @pdurbin ,
Sorry for the delayed response. Looks like the net::ERR_CONTENT_LENGTH_MISMATCH item has been fixed, as I can no longer replicate. This means we're that much closer to integration! 
Really excited to hear that #3657 may be part of the the next sprint. I'm available to test/integrate whatever json you have in mind against the Data Explorer.

**Comment by pdurbin** [comment]
@kevinworthington that's great new that you can't replicate the net::ERR_CONTENT_LENGTH_MISMATCH issue anymore. I think once we get Modular Explore support in #3657 it will be pretty trivial for you to write a JSON file as an "external tool manifest" for Data Explore. The thinking at the moment is that we will list all external tools in the Installation Guide so definition of done for this issue will include adding the manifest there and writing a little blurb about what it does.

It looks like #3657 will be worked on soonish. As I was saying in IRC, I could use some help making the Dataverse code smarter to handle a more complex manifest like Data Explorer requires. If you think you can help with this effort, please take a look at the hack I committed at 272b4c2 to demonstrate the direction the code needs to go. Hopefully the tests are useful at least.

**Comment by kevinworthington** [comment]
Hi @pdurbin ,
The tests are really helpful. I agree that the replaceAll implementation is a bit clumsy and can offer a mechanism to comb over the json looking for dynamic items and replacing them accordingly. Sound good?

**Comment by pdurbin** [comment]
@kevinworthington sure, I'm open to ideas. Thanks!

**Comment by scolapasta** [comment]
@kevinworthington @pdurbin FYI, I was already planning on writing this parser during next sprint. The tests will be useful to make sure it works and I'll likely have some questions on how Data Explorer's APIs.

**Comment by pdurbin** [comment]
@kevinworthington heads up that we had a sprint planning meeting today and Modular Explore (#3657) didn't make it into the next sprint (11/27-12/6) so if you feel like hacking on a parser, please feel free!

**Comment by kevinworthington** [comment]
Hi @pdurbin ,
That's too bad. Last week I submitted a pull request with a regular expression used to parse out dynamic elements from the json string. Did you see that?

**Comment by pdurbin** [comment]
@kevinworthington are you talking about https://github.com/kevinworthington/dataverse/commit/2e749bc0a67520b93977bb7a61b408bd5c12c4fa ? No, I didn't know you had written any code (thanks!) and I don't see a pull request. I found that commit via https://github.com/kevinworthington.atom . Do you want us to take a look at it? Or did you mean some other commit(s)?

**Comment by kevinworthington** [comment]
Hi @pdurbin , Thanks for finding this. I thought the pull request triggered automatically as I edited directly in github. I will create a pull request now so you can take a look.

**Comment by pdurbin** [comment]
@kevinworthington thanks! I mentioned your pull request #4313 at standup this morning and just dragged it to Code Review at https://waffle.io/IQSS/dataverse

**Comment by pdurbin** [comment]
I just pushed some code for #3657 with Data Explorer in mind at 5ae8418

@kevinworthington please see the description I added for Data Explorer and let me know what you'd like it to be for real. Thanks.

**Comment by pdurbin** [comment]
@kevinworthington thanks for improving the description for Data Explorer in 488b604

A couple things. I left a longish comment about popups at https://github.com/IQSS/dataverse/issues/3657#issuecomment-355619954 that I'd like you and others interested in Data Explorer to check out.

Also, I just opened https://github.com/scholarsportal/Dataverse-Data-Explorer/issues/2 about CORS but the larger issue is documentation for Data Explorer. I'm not sure if we'll try to QA this issue (#4249) along with #3657 (when it's ready) or not but for QA of Data Explorer to happen, the Data Explorer README or whatever you'd like us to link to should be ready for QA. Please just indicate here when you feel like the Data Explorer documentation is ready. I suspect that that JSON "manifest" will need to change so I'm sure the README will be partially blocked on #3657 but I wanted to give you a heads up that this freight train is starting to move, despite all the snow. 😄 🚆 🚋 🚄 ❄️

**Comment by pdurbin** [comment]
It looks like the CORS thing was a red herring. Thanks for helping me come up with a fix in e873c6e @kevinworthington . Still let's continue to think about what it will take before this issue is ready for QA. What the "definition of done" for this issue is, I mean.

**Comment by kevinworthington** [comment]
Hi @pdurbin,
Glad to have helped. It's really exciting to see your dev instance linking to the Data Explorer. For us, done is being able to connect the Data Explorer following the prescribed workflow without having to modify any code beyond the config file.
I have updated the README file for the Data Explorer and will add further instructions as they become available.

**Comment by pdurbin** [comment]
@kevinworthington cool, your README is looking good. Heads up in the 233cdc1 I changed the external tool manifest format to require a "type" such as "explore" or "configure".

**Comment by kevinworthington** [comment]
Thanks for the manifest update @pdurbin. Seems like a very useful parameter to have incorporated.

**Comment by kevinworthington** [comment]
Hi @pdurbin,
Thanks for reporting those issues and offering the option to have the Data Explorer featured along with other external tools. I've made a note to myself to bring this up during our call today.

**Comment by pdurbin** [comment]
@kevinworthington nice chatting with you and @amberleahey this afternoon. As we discussed, pull request #4426 was just merged and I'm hoping you can make a pull request for the following:

- Delete `doc/sphinx-guides/source/_static/installation/files/root/external-tools/dataExplorer.json` (As the external tool author, you'll decide where this "manifest" is hosted. #4429 is about moving TwoRavens to this model now that it's an external tool.)
- Edit `doc/sphinx-guides/source/installation/external-tools.rst` to replace the mention of this issue (#4249) with a URL that has instructions on what sysadmins should do to enable Data Explorer in their installation of Dataverse.

Thanks!

**Comment by kevinworthington** [comment]
Hi @pdurbin,
Thanks for clearly outlining the steps needed to assist with this integration piece. I'll be sure to have it complete before the community call on Tuesday :)

PS: It was nice chatting with you too.

**Comment by pdurbin** [comment]
@kevinworthington thanks for making pull request #4443 which is exactly what I asked for. I approved the pull request through the GitHub GUI but I moved this issue to code review at https://waffle.io/IQSS/dataverse in case any other developers want to take a look. I'm sure we'll touch on it during standup tomorrow. Thanks!

**Comment by djbrooke** [comment]
Thanks @kevinworthington!
